pkg/services/frontend/index.html

```diff
@@ -56,46 +161,104 @@
       [[end]]
+    </script>
 
-      window.__grafana_load_failed = function(...args) {
-        console.error('Failed to load Grafana', ...args);
-      };
+    <script nonce="[[.Nonce]]">
+      // Wrap in an IIFE to avoid polluting the global scope. Intentionally global-scope properties
+      // are explicitly assigned to the `window` object.
+      (() => {
+        window.__grafana_load_failed = function(...args) {
+          console.error('Failed to load Grafana', ...args);
+          document.querySelector('.fs-variant-loader').classList.add('fs-hidden');
+          document.querySelector('.fs-variant-error').classList.remove('fs-hidden');
+        };
 
-      window.__grafana_boot_data_promise = new Promise(async (resolve) => {
-        const bootData = await fetch("/bootdata");
+        window.onload = function() {
+          if (window.__grafana_app_bundle_loaded) {
+            return;
+          }
+          window.__grafana_load_failed();
+        };
 
-        const rawBootData = await bootData.json();
+        let hasSetLoading = false;
+        function setLoading() {
+          if (hasSetLoading) {
+            return;
+          }
 
-        window.grafanaBootData = {
-          _femt: true,
-          ...rawBootData,
+          document.querySelector('.fs-loader').classList.add('fs-loader-starting-up');
+          hasSetLoading = true;
         }
 
-        // The per-theme CSS still contains some global styles needed
-        // to render the page correctly.
-        const cssLink = document.createElement("link");
-        cssLink.rel = 'stylesheet';
+        const CHECK_INTERVAL = 1 * 1000;
+
+        async function loadBootData() {
+          while (true) {
+            const resp = await fetch("/bootdata");
+
+            const textResponse = await resp.text();
+            let rawBootData;
+            try {
+              rawBootData = JSON.parse(textResponse);
+            } catch {
+              // If we haven't gotten a JSON response, there must be an error
+              throw new Error("Unexpected response type: " + textResponse);
+            }
+
+            if (resp.status === 503 && rawBootData.code === 'Loading') {
+              // If the response is 503, poll until we get a valid response.
+              setLoading();
+              await new Promise(resolve => setTimeout(resolve, CHECK_INTERVAL));
+              continue;
+            } else if (!resp.ok) {
+              // Or if we got a non-200 response, throw an error.
+              const errorText = await resp.text();
+              throw new Error("Unexpected response body: " + errorText);
+            }
 
-        let theme = window.grafanaBootData.user.theme;
-        if (theme === "system") {
-          const darkQuery = window.matchMedia("(prefers-color-scheme: dark)");
-          theme = darkQuery.matches ? 'dark' : 'light';
+            return rawBootData;
+          }
```

Comment:
i... don't understand this bit 😬 `while (true) {}` feels all sorts of wrong 😅 
what's the intention of this code? i think it's to poll every second, ignoring 503 responses, until we either get a good response (200 with valid json) or a bad response (anything else). please correct me if that's wrong!

if so, wouldn't this be simpler as a single function passed to `setInterval`?

Reply:
we should probably also define a timeout here e.g. if it's still 503ing after 2 minutes, throw a timeout error

Reply:
Heh, fair enough. This kind of pattern comes naturally to me - an infinite loop with an async wait in it, but I understand your concern. I'll re-jigger it.